eth/tracers/api_blockResult.go

```diff
@@ -0,0 +1,347 @@
+		block, err = api.blockByNumber(ctx, number)
+	}
+	if hash, ok := blockNrOrHash.Hash(); ok {
+		block, err = api.blockByHash(ctx, hash)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if block.NumberU64() == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+	if config == nil {
+		config = &TraceConfig{
+			LogConfig: &vm.LogConfig{
+				EnableMemory:     false,
+				EnableReturnData: true,
+			},
+		}
+	} else if config.Tracer != nil {
+		config.Tracer = nil
+		log.Warn("Tracer params is unsupported")
+	}
+
+	// create current execution environment.
+	env, err := api.createTraceEnv(ctx, config, block)
+	if err != nil {
+		return nil, err
+	}
+
+	return api.getBlockResult(block, env)
+}
+
+// Make trace environment for current block.
+func (api *API) createTraceEnv(ctx context.Context, config *TraceConfig, block *types.Block) (*traceEnv, error) {
+	parent, err := api.blockByNumberAndHash(ctx, rpc.BlockNumber(block.NumberU64()-1), block.ParentHash())
+	if err != nil {
+		return nil, err
+	}
+	reexec := defaultTraceReexec
+	if config != nil && config.Reexec != nil {
+		reexec = *config.Reexec
+	}
+	statedb, err := api.backend.StateAtBlock(ctx, parent, reexec, nil, true, true)
+	if err != nil {
+		return nil, err
+	}
+
+	env := &traceEnv{
+		config:   config,
+		signer:   types.MakeSigner(api.backend.ChainConfig(), block.Number()),
+		state:    statedb,
+		tracer:   vm.NewStructLogger(config.LogConfig),
+		blockCtx: core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil),
+		StorageTrace: &types.StorageTrace{
+			RootBefore:    parent.Root(),
+			RootAfter:     block.Root(),
+			Proofs:        make(map[string][]hexutil.Bytes),
+			StorageProofs: make(map[string]map[string][]hexutil.Bytes),
+		},
+		executionResults: make([]*types.ExecutionResult, block.Transactions().Len()),
+	}
+
+	for _, coinbase := range []common.Address{block.Coinbase(), api.backend.Coinbase()} {
+		if coinbase == (common.Address{}) {
+			continue
+		}
+		key := coinbase.String()
+		if _, exist := env.Proofs[key]; !exist {
+			proof, err := env.state.GetProof(coinbase)
+			if err != nil {
+				log.Error("Proof for coinbase not available", "coinbase", coinbase, "error", err)
+				// but we still mark the proofs map with nil array
+			}
+			wrappedProof := make([]hexutil.Bytes, len(proof))
+			for i, bt := range proof {
+				wrappedProof[i] = bt
+			}
+			env.Proofs[key] = wrappedProof
+		}
+	}
+	return env, nil
+}
+
+func (api *API) getBlockResult(block *types.Block, env *traceEnv) (*types.BlockResult, error) {
+	// Execute all the transaction contained within the block concurrently
+	var (
+		txs  = block.Transactions()
+		pend = new(sync.WaitGroup)
+		jobs = make(chan *txTraceTask, len(txs))
+	)
+	threads := runtime.NumCPU()
+	if threads > len(txs) {
+		threads = len(txs)
+	}
+	for th := 0; th < threads; th++ {
+		pend.Add(1)
+		go func() {
+			defer pend.Done()
+			// Fetch and execute the next transaction trace tasks
+			for task := range jobs {
```

Comment:
how do we know whether we have loop along all the txs?

Reply:
I don't think we can loop over all the jobs here, for example:

```
package main

import (
	"fmt"
	"sync"
)

func main() {
	jobs := make(chan int, 10)

	wg := new(sync.WaitGroup)

	wg.Add(1)
	go func() {
		defer wg.Done()
		for j := range jobs {
			fmt.Println(j)
		}
	}()

	jobs <- 1
}
```

This code exit without waiting

Reply:
The loop will finish when `jobs` is closed.